### PR TITLE
[DataCacheCore] Make players store the current playing item in data cache core, remove CApp dependency

### PR DESCRIPTION
**Pull request by enen92**
## Description
Looking into how to decouple components from GUI and from the Application I've found that a lot of examples just call into g_application to obtain either the item that is being played or the player times (max and current playing time). In fact I think that the item being played is something that could be cached elsewhere (e.g. exploiting data cache core) along side other information about the running player that's already available there for the info interface to consume.
Doing this I was able to remove the `Application.h` include from the UPnP Renderer. Next step will be dropping the ApplicationPlayer dependencies.

This "depends" on https://github.com/xbmc/xbmc/pull/24402
Note: clang-format is broken on upnp renderer file, I'll prefer to do it in bulk later than now to avoid the exploding diff.

## Motivation and context
Untangle stuff, being able to run headless some day

## How has this been tested?
Runtime tested on linux with breakpoints in the affected places to confirm the data that was returned previously and the one that is return now match completely

## What is the effect on users?
None, internal refactor/cleanup.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [x] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by enen92** [comment]
@garbear are you ok with the change now? Also https://github.com/xbmc/xbmc/pull/24402

File: xbmc/cores/DataCacheCore.h [hunk 1]
```diff
@@ -195,8 +198,23 @@ class CDataCacheCore
...
+   * \param item - the item to store
+   */
+  void SetFileItem(const CFileItem& item);
+
+  /*!
+   * \brief Get the current playing item
+   *
+   * \return the item currently cached by the player
+   */
+  const std::shared_ptr<CFileItem> GetFileItem() const;
```
Comment (neo1973) [inline xbmc/cores/DataCacheCore.h:213]: The `const` on the `std::shared_ptr` doesn't make sense here, it's returned by value so the `const` doesn't have an effect. What you probably want is `std::shared_ptr<const CFileItem> GetFileItem() const;` but that requires const-correctness fixes in a number of files.

Comment (howie-f) [reply xbmc/cores/DataCacheCore.h:213]: could be the intention is to return a const reference to the shared_ptr. in general we should not return plain const of any type

Comment (enen92) [reply xbmc/cores/DataCacheCore.h:213]: Can't be a const ref as the item might be a nullptr if not set. I'll drop the const qualifier, it doesn't make sense on return.

**Comment by enen92** [comment]
Feedback on this one is also appreciated

**Review by neo1973 — APPROVED** [review]

File: xbmc/network/upnp/UPnPRenderer.cpp [hunk 2]
```diff
@@ -251,8 +251,12 @@ void CUPnPRenderer::Announce(ANNOUNCEMENT::AnnouncementFlag flag,
 
     if (message == "OnPlay" || message == "OnResume")
     {
-      avt->SetStateVariable("AVTransportURI", g_application.CurrentFile().c_str());
-      avt->SetStateVariable("CurrentTrackURI", g_application.CurrentFile().c_str());
+      const std::shared_ptr<CFileItem> item = CServiceBroker::GetDataCacheCore().GetFileItem();
+      if (item)
+      {
+                avt->SetStateVariable("AVTransportURI", item->GetPath().c_str());
+                avt->SetStateVariable("CurrentTrackURI", item->GetPath().c_str());
```
Comment (ksooo) [inline xbmc/network/upnp/UPnPRenderer.cpp:258]: Indentation looks wrong.

**Review by ksooo — COMMENTED** [review]
With this change, you are making CDataCacheCore dependent on CFileItem. So far, this has been a rather low level class, only carrying basic simple data structures. A CFileItem is rather heavy weight.

Looking at what you do with the item, it would be sufficient to store playing item's path in CDataCacheCore.

Just my 2ct.

File: xbmc/network/upnp/UPnPRenderer.cpp (hunk 2 above)
Comment (enen92) [reply xbmc/network/upnp/UPnPRenderer.cpp:256]: I'll fix that later in bulk, the file is completely fucked up

Comment (ksooo) [reply xbmc/network/upnp/UPnPRenderer.cpp:256]: alright

**Comment by garbear** [comment]
Is upnp player the only place where `GetFileItem()` is called? It looks like you're swapping out one global for another, though given that the global being swapped out is CApp this seems like a good trade.

What if the fileitem changes? Is it a problem if datacachecore has a stale fileitem or stale time?

**Comment by enen92** [comment]
> Is upnp player the only place where `GetFileItem()` is called? It looks like you're swapping out one global for another, though given that the global being swapped out is CApp this seems like a good trade.
> 
> What if the fileitem changes? Is it a problem if datacachecore has a stale fileitem or stale time?

Currently yes but my goal is to evaluate the replacement of more usages of capp.GetCurrentFile() in other places (it has to start somewhere, and from experience it's better to start with unmaintained and somehow contained..legacy components like UPnP). When a fileitem reaches the player it must be always fully resolved which includes not only the paths but any art associated with the item. Part of this art comes from the infotags so that should also be filled correctly.
There's no reason for the player to change anything on the item during playback, the only properties that may change are the play count or resume information. Those are set when the player closes the operation and are persisted to the dB anyway (so the next time the player is started the item already includes them). Furthermore, since datacachecore only stores the item during the lifetime of the player, when the player closes it resets the item cached.
Long story short, having stale info on the item is a risk but if somehow that occurs imoh the root cause of the issue is elsewhere

**Comment by garbear** [comment]
> There's no reason for the player to change anything on the item during playback

While true for most player cores, RetroPlayer uses RetroAchievements.org for rich presence captioning, and RA.org provides game metadata that gets synced to the three places where the current fileitem lives: In the player, in the guiinfo manager, and in CApp. That's why I'm hesitant here to add a fourth location, but if it removes any of the other three then it's a beneficial change. We could argue that RA.org metadata should be set outside of RetroPlayer, but that's what it currently does.

**Comment by enen92** [comment]
> > There's no reason for the player to change anything on the item during playback
> 
> While true for most player cores, RetroPlayer uses RetroAchievements.org for rich presence captioning, and RA.org provides game metadata that gets synced to the three places where the current fileitem lives: In the player, in the guiinfo manager, and in CApp. That's why I'm hesitant here to add a fourth location, but if it removes any of the other three then it's a beneficial change. We could argue that RA.org metadata should be set outside of RetroPlayer, but that's what it currently does.

Thanks for the detailed explanation, this is something I lack knowledge about.
I've taken a look of current usages of:
- `g_application.CurrentFile()` -> all usages (except upnp which is being nuked here) happen on window subtitles and video bookmarks to obtain the item from the player. It can be replaced with datacachecore with ease. Calls to `ResetCurrentItem()` happens as callbacks to thread messages related to player operations too.
- `GUIInfomanager` -> all usages except PVR seem to simply use it to get the item that is being played. PVR seems to do the same too but needs further investigation (not familiar with CPVRGUIChannelNavigator). In RP you use `TMSG_UPDATE_PLAYER_ITEM` which could be replaced with datacachecore too.

So personally I think the introduction of the fileitem in datacache core can lead to the removal of the other two which are high tight points between components and GUI/CApp. I am happy to have a go at it for v22, for v21 nuking the other usages is too risky at this point.

**Review by garbear — APPROVED** [review]
I'm OK with this change then, and I think that in v22, moving the other fileitems to datacachecore is a good idea.

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
> I'm OK with this change then, and I think that in v22, moving the other fileitems to datacachecore is a good idea.

We've to start somewhere. Unfortunately the more I dive into the fileitem implementation the more dependencies I found on GUI :')

**Comment by enen92** [comment]
I'll move this one to v22, something tells me I'll come back to this later.

**Comment by garbear** [comment]
@enen92 Now that we're in v22, wanna rebase and merge?

**Comment by garbear** [comment]
Is this wanted for Beta 2?
